// @project-serum/anchor/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wallet = exports.utils = exports.web3 = exports.BN = exports.Provider = exports.getProvider = exports.setProvider = exports.AccountsCoder = exports.TypesCoder = exports.StateCoder = exports.EventCoder = exports.InstructionCoder = exports.Coder = exports.StateClient = exports.AccountClient = exports.Program = exports.workspace = void 0;
const bn_js_1 = __importDefault(require("bn.js"));
exports.BN = bn_js_1.default;
const web3 = __importStar(require("@solana/web3.js"));
exports.web3 = web3;
const provider_1 = __importStar(require("./provider"));
exports.Provider = provider_1.default;
Object.defineProperty(exports, "getProvider", { enumerable: true, get: function () { return provider_1.getProvider; } });
Object.defineProperty(exports, "setProvider", { enumerable: true, get: function () { return provider_1.setProvider; } });
Object.defineProperty(exports, "Wallet", { enumerable: true, get: function () { return provider_1.NodeWallet; } });
const coder_1 = __importStar(require("./coder"));
exports.Coder = coder_1.default;
Object.defineProperty(exports, "InstructionCoder", { enumerable: true, get: function () { return coder_1.InstructionCoder; } });
Object.defineProperty(exports, "EventCoder", { enumerable: true, get: function () { return coder_1.EventCoder; } });
Object.defineProperty(exports, "StateCoder", { enumerable: true, get: function () { return coder_1.StateCoder; } });
Object.defineProperty(exports, "TypesCoder", { enumerable: true, get: function () { return coder_1.TypesCoder; } });
Object.defineProperty(exports, "AccountsCoder", { enumerable: true, get: function () { return coder_1.AccountsCoder; } });
const workspace_1 = __importDefault(require("./workspace"));
exports.workspace = workspace_1.default;
const utils = __importStar(require("./utils"));
exports.utils = utils;
const program_1 = require("./program");
Object.defineProperty(exports, "Program", { enumerable: true, get: function () { return program_1.Program; } });
const namespace_1 = require("./program/namespace");
Object.defineProperty(exports, "AccountClient", { enumerable: true, get: function () { return namespace_1.AccountClient; } });
Object.defineProperty(exports, "StateClient", { enumerable: true, get: function () { return namespace_1.StateClient; } });
